jobs/pbm_helpers.go

```diff
@@ -0,0 +1,378 @@
+	Name     string `json:"name"`
+	Error    string `json:"error"`
+}
+
+type pbmStatus struct {
+	Backups struct {
+		Type       string        `json:"type"`
+		Path       string        `json:"path"`
+		Region     string        `json:"region"`
+		Snapshot   []pbmSnapshot `json:"snapshot"`
+		PitrChunks struct {
+			Size int `json:"size"`
+		} `json:"pitrChunks"`
+	} `json:"backups"`
+	Cluster []struct {
+		Rs    string `json:"rs"`
+		Nodes []struct {
+			Host  string `json:"host"`
+			Agent string `json:"agent"`
+			Ok    bool   `json:"ok"`
+		} `json:"nodes"`
+	} `json:"cluster"`
+	Pitr struct {
+		Conf bool `json:"conf"`
+		Run  bool `json:"run"`
+	} `json:"pitr"`
+	Running struct {
+		Type    string `json:"type"`
+		Name    string `json:"name"`
+		StartTS int    `json:"startTS"`
+		Status  string `json:"status"`
+		OpID    string `json:"opID"`
+	} `json:"running"`
+}
+
+func execPBMCommand(ctx context.Context, dbURL *url.URL, to interface{}, args ...string) error {
+	nCtx, cancel := context.WithTimeout(ctx, cmdTimeout)
+	defer cancel()
+
+	args = append(args, "--out=json", "--mongodb-uri="+dbURL.String())
+	cmd := exec.CommandContext(nCtx, pbmBin, args...) // #nosec G204
+
+	b, err := cmd.Output()
+	if err != nil {
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) {
+			return errors.New(string(exitErr.Stderr))
+		}
+		return err
+	}
+
+	return json.Unmarshal(b, to)
+}
+
+func pbmSetupS3(ctx context.Context, l logrus.FieldLogger, dbURL *url.URL, prefix string, s3Config *S3LocationConfig, resync bool) error {
+	l.Info("Configuring S3 location.")
+	nCtx, cancel := context.WithTimeout(ctx, cmdTimeout)
+	defer cancel()
+
+	confFile, err := writePBMConfigFile(prefix, s3Config)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer os.Remove(confFile) //nolint:errcheck
+
+	output, err := exec.CommandContext( //nolint:gosec
+		nCtx,
+		pbmBin,
+		"config",
+		"--mongodb-uri="+dbURL.String(),
+		"--file="+confFile,
+	).CombinedOutput()
+
+	if err != nil {
+		return errors.Wrapf(err, "pbm config error: %s", string(output))
+	}
+
+	if resync {
+		nCtx, cancel := context.WithTimeout(ctx, cmdTimeout)
+		defer cancel()
+
+		output, err = exec.CommandContext( //nolint:gosec
+			nCtx,
+			pbmBin,
+			"config",
+			"--mongodb-uri="+dbURL.String(),
+			"--force-resync",
+		).CombinedOutput()
+
+		if err != nil {
+			return errors.Wrapf(err, "pbm config error: %s", string(output))
+		}
+	}
+
+	return nil
+}
+
+func retrieveLogs(ctx context.Context, dbURL *url.URL, event string) ([]pbmLogEntry, error) {
+	var logs []pbmLogEntry
+
+	if err := execPBMCommand(ctx, dbURL, &logs, "logs", "--event="+event, "--tail=0"); err != nil {
+		return nil, err
+	}
+
+	return logs, nil
+}
+
+type pbmStatusCondition func(s pbmStatus) (bool, error)
+
+func pbmNoRunningOperations(s pbmStatus) (bool, error) {
+	return s.Running.Status == "", nil
+}
+
+func pbmBackupFinished(name string) pbmStatusCondition {
+	started := false
+	snapshotStarted := false
```

Comment:
There is bug in pbm 1.6.0, when first state is `error` and then `starting` -> ... `done`, so it's to prevent from catching first wrong `error` state
Fixed in [master](https://github.com/percona/percona-backup-mongodb/commit/22fd155e0463989bf2756968d51412a304bd398d#diff-ce0c03299a55f868b3214510943a2bab26a27885a6f5877d6d5aaa4391437ecdR146-R148)